File: object-oriented-programming/fundamentals/visual-basic/deck_class.vb
' Raspberry Pi Foundation
' Usage licensed under CC BY-NC-SA 4.0
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program


Class PlayingCard
    Private suit As String
    Private rank As String
    Private value As Integer

    ' Constructor method
    Public Sub New(ByVal givenSuit As String, ByVal givenRank As String, ByVal givenValue As Integer)
        suit = givenSuit
        rank = givenRank
        value = givenValue
    End Sub
End Class


Class Deck

    Private cards As PlayingCard() = New PlayingCard(51) {} ' Declare an array of 52 elements
    Private suits As String() = {"Clubs", "Diamonds", "Hearts", "Spades"}
    Private ranks As String() = {"Ace", "2", "3", "4", "5", "6", "7", "8", "9", "10", "Jack", "Queen", "King"}
    Private values As Integer() = {11, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10}

    ' Constructor method
    Public Sub New()
        Dim cardNumber As Integer = 0

        ' Repeat for all 4 suits
        For i As Integer = 0 To suits.Length - 1
            ' Repeat for all 13 ranks of a suit
            For j As Integer = 0 To ranks.Length - 1
                Dim newCard As PlayingCard = New PlayingCard(suits(i), ranks(j), values(j))
                cards(cardNumber) = newCard
                cardNumber += 1
            Next
        Next

        ' Show that all 52 playing cards have been initialised
        Console.WriteLine($"Initialised {cardNumber} playing cards")
    End Sub

End Class


Module Program

    ' The Main method is the default entry point for all VB programs
    Sub Main()
        ' Instantiate a new deck object
        Dim myDeck As Deck = New Deck()
    End Sub

End Module

File: object-oriented-programming/fundamentals/c-sharp/deck_class.cs
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science 
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;

namespace AdaCodeSamples {
    class PlayingCard {
        private string suit;
        public string rank;
        private int value;

        // Constructor method
        public PlayingCard(string givenSuit, string givenRank, int givenValue) {
            suit = givenSuit;
            rank = givenRank;
            value = givenValue;
        }
    }
    

    class Deck {
        private PlayingCard[] cards = new PlayingCard[52]; // Declare an array of 52 elements
        private string[] suits = { "Clubs", "Diamonds", "Hearts", "Spades" };
        private string[] ranks = { "Ace", "2", "3", "4", "5", "6", "7", "8", "9", "10", "Jack", "Queen", "King" };
        private int[] values = { 11, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10 };

        // Constructor method
        public Deck() {
            int cardNumber = 0;

            // Repeat for all 4 suits
            for (int i = 0; i < suits.Length; i++) {
                // Repeat for all 13 ranks of a suit
                for (int j = 0; j < ranks.Length; j++) {
                    PlayingCard newCard = new PlayingCard(suits[i], ranks[j], values[j]);
                    cards[cardNumber] = newCard;
                    cardNumber += 1;
                }
            }

            // Show that all 52 playing cards have been initialised
            Console.WriteLine($"Initialised {cardNumber} playing cards");
        }
        

    }


    class Testing {
        // The Main method is the entry point for all C# programs
        public static void Main() {
            // Instantiate a new deck object
            Deck myDeck = new Deck();
        }


    }
}